### ISSUE TITLE
[SFT] On Backpropagation of Nan Loss and Data Truncation

### ISSUE THREAD

**Issue by cpchenpi** [volcengine/verl#785]
When using `fsdp_sft_trainer.py` for training, it was observed that the model would produce NaN loss at a fixed iteration and subsequently crash, degrading into repeatedly generating '!!!!!!'.

Upon investigation, it was found that this was caused by data truncation, which resulted in the entire batch lacking positions with loss mask. Consequently, a division by zero occurred during calculating the average, causing the loss to become NaN, and the model crashed after backpropagation. This issue is particularly common when the truncation length is short, and the prompt length may be relatively long. There are multiple ways to solve this problem, such as removing data with long prompts or directly preventing division by zero.

https://github.com/volcengine/verl/blob/a1dd9224b657a0ff7438b0a54e67bb847eb6dafb/verl/trainer/fsdp_sft_trainer.py#L369-L377

I noticed several issues also reported NaN loss and model crashes, e.g. [Issue 747](https://github.com/volcengine/verl/issues/747). Although these did not occur during SFT, I suspect the mechanism causing the problem is similar.

**Comment by hiyouga** [comment]
Closed by #786
